keps/0024-parameter-enhancement.md

```diff
@@ -0,0 +1,154 @@
+    - [Notes](#notes)
+  - [Alternatives](#alternatives)
+    - [Convert YAML in string to dictionaries](#convert-yaml-in-string-to-dictionaries)
+    - [Allow Arbitrary Parameter Values](#allow-arbitrary-parameter-values)
+  - [Implementation History](#implementation-history)
+
+## Summary
+
+This KEP aims to improve the use-cases that can be covered by parameters. By supporting additional types, templates can access lists and dictionaries. Furthermore, validation can be implemented for known types.
+
+## Motivation
+
+Currently only string parameters are supported by KUDO. This limits the possible use-cases of parameters. Various use cases benefit from having parameters that are lists or dictionaries. While some of this functionality can be implemented in Go templates using [Sprig functions][1], this is unintuitive.
+Having parameter types creates the possibility for parameter validation.
+
+### Goals
+
+- Provide operator developers with more versatile parameters
+- Allow the definition of validation functions for parameters
+
+### Non-Goals
+
+- Change existing parameter definitions in `params.yaml`
+
+## Proposal
+
+### Support Optional Parameter Types
+
+By adding an optional `type` field to the list of parameters, parameter types can be specified. If this field isn't present, it defaults to the `string` type. Other types like `list`, `dict` or `integer` are possible. KUDO's renderer uses the type information to convert the string value to the respective type. For example, for the `dict` type, the input value would be YAML that is unmarshalled to a `map[string]interface{}`. Go's template engine allows to access and iterate list and dictionary items.
+
+Consider a cluster that has nodes in multiple regions. We want to ensure that a nginx deployment has 2 replicas in one region and 1 replica in the other one. Furthermore they should support both HTTP as well as HTTPS, hence need a list of port definitions. These parameters are described by `list` and `dict` types:
+
+```yaml
+parameters:
+  - name: ports
+    description: Ports
+    type: list
+    default: [ 80, 443 ]
+  - name: topology
+    description: Node topology
+    type: dict
+    default:
+      - region: us-east-1:
+        replicas: 2
+      - region: us-west-1:
+        replicas: 1
+```
+
+The deployment template then iterates over the items of the respective parameters:
+
+```yaml
+{{ $ports := .Params.ports }}
+{{ range .Params.topology }}
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+  labels:
+    app.kubernetes.io/name: nginx
+spec:
+  replicas: {{ .replicas }}
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: nginx
+  template:
+    metadata:
+      name: nginx
+      labels:
+        app.kubernetes.io/name: nginx
+    spec:
+      nodeSelector:
+        topology.kubernetes.io/region: {{ .region }}
+      containers:
+      - name: nginx
+        image: nginx:stable
+        ports:
+        {{ range $ports }}
```

Comment:
```suggestion
        {{ range $.Params.ports }}
```
Let's keep the example validator-compatible, and it's shorter this way too, without sacrificing clarity.

Reply:
:+1:, it's clearer. Though it's still a bug in the validator if it isn't Go template compatible ;)